src/borg/key.py

```diff
@@ -57,6 +59,80 @@ class RepoKeyNotFoundError(Error):
     """No key entry found in the config of repository {}."""
 
 
+class SignatureError(Error):
+    """Invalid signature for {}."""
+
+
+class SignedFile(FileLikeWrapper):
+    # generic enough that it can be used without a KeyBase
+    def __init__(self, key, path, write):
+        self.path = path
+        self.writing = write
+        mode = 'wb' if write else 'rb'
+        self.file_fd = open(path, mode)
+        self.signer = StreamSigner_HMAC_SHA512(key, self.file_fd, write)
+        self.fd = self.signer  # for FileLikeWrapper
+        self.sign_filename()
+        if write:
+            self.signatures = {}
+        else:
+            self.signatures = self.read_signatures(path, self.signer)
+        self.failed = False
+
+    def sign_filename(self):
+        # Sign the name of the file as well, but only the basename, ie. not the path. In Borg
+        # the name itself encodes the context (eg. index.N, cache, files), while the path doesn't matter,
+        # and moving eg. a repository or cache directory is supported.
+        # Changing the name however imbues a change of context that is not permissible.
+        filename = os.path.basename(self.path)
+        self.signer.update(b'%d' % len(filename))
+        self.signer.update(filename.encode())
```

Comment:
IIRC the default was adjustable on Python 2 (derived by locale, even?), but in 3 it's always utf-8. Sane defaults are nice :)